File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.swift
class Solution {
    func longestOnes(_ nums: [Int], _ k: Int) -> Int {
        var l = 0
        var numZeroes = 0
        var currentLongest = 0
        var res = 0
        for r in 0..<nums.count {
            if nums[r] == 0 {
                numZeroes += 1
            }
            currentLongest += 1

            while numZeroes > k {
                if nums[l] == 0 {
                    numZeroes -= 1
                }
                currentLongest -= 1
                l += 1
            }

            res = max(res, currentLongest)
        }

        return res
    }
}

File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        """
        [1,1,1,0,0,0,1,1,1,1,0]
        max consecutive number of 1s: 4
        k = 2
        at least 6, might be more if its connected (just count the number of consecutive ones again)


        Test run
        [1,1,1,0,0,0,1,1,1,1,0]
                   l
                               r

         numZeroes = 2
         res = 6
         currLongest = 6
        """
        l = 0
        numZeros = 0
        res = 0
        currLongest = 0
        for r in range(len(nums)):
            if nums[r] == 0:
                numZeros += 1
            currLongest += 1 
                        
            while numZeros > k:
                if nums[l] == 0:
                    numZeros -= 1
                currLongest -= 1
                l += 1
            
            res = max(res, currLongest)
        
        return res

            



